lemmy_db/src/post_report.rs

```diff
@@ -0,0 +1,195 @@
+        post_creator_name -> Varchar,
+    }
+}
+
+#[derive(Identifiable, Queryable, Associations, PartialEq, Serialize, Deserialize, Debug)]
+#[belongs_to(Post)]
+#[table_name = "post_report"]
+pub struct PostReport {
+  pub id: i32,
+  pub creator_id: i32,
+  pub post_id: i32,
+  pub post_name: String,
+  pub post_url: Option<String>,
+  pub post_body: Option<String>,
+  pub reason: String,
+  pub resolved: bool,
+  pub resolver_id: Option<i32>,
+  pub published: chrono::NaiveDateTime,
+  pub updated: Option<chrono::NaiveDateTime>,
+}
+
+#[derive(Insertable, AsChangeset, Clone)]
+#[table_name = "post_report"]
+pub struct PostReportForm {
+  pub creator_id: i32,
+  pub post_id: i32,
+  pub post_name: String,
+  pub post_url: Option<String>,
+  pub post_body: Option<String>,
+  pub reason: String,
+}
+
+impl Reportable<PostReportForm> for PostReport {
+  fn report(conn: &PgConnection, post_report_form: &PostReportForm) -> Result<Self, Error> {
+    use crate::schema::post_report::dsl::*;
+    insert_into(post_report)
+      .values(post_report_form)
+      .get_result::<Self>(conn)
+  }
+
+  fn resolve(conn: &PgConnection, report_id: i32, by_user_id: i32) -> Result<usize, Error> {
+    use crate::schema::post_report::dsl::*;
+    update(post_report.find(report_id))
+      .set((
+        resolved.eq(true),
+        resolver_id.eq(by_user_id),
+        updated.eq(naive_now()),
+      ))
+      .execute(conn)
+  }
+
+  fn unresolve(conn: &PgConnection, report_id: i32) -> Result<usize, Error> {
+    use crate::schema::post_report::dsl::*;
+    update(post_report.find(report_id))
+      .set((
+        resolved.eq(false),
+        updated.eq(naive_now()),
+      ))
+      .execute(conn)
+  }
+}
+
+#[derive(
+Queryable, Identifiable, PartialEq, Debug, Serialize, Deserialize, Clone,
+)]
+#[table_name = "post_report_view"]
+pub struct PostReportView {
+  pub id: i32,
+  pub creator_id: i32,
+  pub post_id: i32,
+  pub post_name: String,
+  pub post_url: Option<String>,
+  pub post_body: Option<String>,
+  pub reason: String,
+  pub resolved: bool,
+  pub resolver_id: Option<i32>,
+  pub published: chrono::NaiveDateTime,
+  pub updated: Option<chrono::NaiveDateTime>,
+  pub current_post_name: String,
+  pub current_post_url: Option<String>,
+  pub current_post_body: Option<String>,
+  pub community_id: i32,
+  pub creator_name: String,
+  pub post_creator_id: i32,
+  pub post_creator_name: String,
+}
+
+impl PostReportView {
+  pub fn read(conn: &PgConnection, report_id: i32) -> Result<Self, Error> {
+    use super::post_report::post_report_view::dsl::*;
+    post_report_view
+      .find(report_id)
+      .first::<Self>(conn)
+  }
+
+  pub fn get_report_count(conn: &PgConnection, community_ids: &Vec<i32>) -> Result<i32, Error> {
+    use super::post_report::post_report_view::dsl::*;
+    post_report_view
+      .filter(resolved.eq(false).and(community_id.eq_any(community_ids)))
+      .select(sql::<sql_types::Integer>("COUNT(*)"))
```

Comment:
This seems pretty strange... it looks like diesel has a count expression: https://docs.diesel.rs/1.4.x/diesel/dsl/fn.count.html

Reply:
word, their docs were down when i wrote that so i pulled it from stack overflow lmao

Reply:
oh, actually, i see why you'd do it this way. `count` will return `sql_types::BigInt` so to avoid that you just use raw sql and cast it to an integer

Reply:
Using the diesel count function looks cleaner so I would definitely go with that. You can cast the return value to i32 if needed.